internal/datadog/metrics_client.go

```diff
@@ -0,0 +1,99 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package datadog // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/datadog"
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+type metricsClient struct {
+	meter  metric.Meter
+	gauges map[string]float64
+	mutex  sync.Mutex
+}
+
+var mutex sync.Mutex
+
+// NewMetricClient returns a new metrics client
+func NewMetricClient(meter metric.Meter) metrics.StatsClient {
+	// We need to lock the mutex here to avoid
+	// duplicate metrics being created by both datadogconnector and exporter
+	mutex.Lock()
+	defer mutex.Unlock()
+	m := &metricsClient{
+		meter:  meter,
+		gauges: make(map[string]float64),
+	}
+	metrics.Client = m
+	return m
+}
+
+func (m *metricsClient) Gauge(name string, value float64, tags []string, _ float64) error {
+	// The last parameter is rate, but we're omitting it because rate does not have effect for gauge points: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/dedd44436ae064f5a0b43769d24adf897533957b/receiver/statsdreceiver/internal/protocol/metric_translator.go#L153-L156
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	if _, ok := m.gauges[name]; ok {
+		m.gauges[name] = value
+		return nil
+	}
+	m.gauges[name] = value
+	_, err := m.meter.Float64ObservableGauge(name, metric.WithFloat64Callback(func(_ context.Context, f metric.Float64Observer) error {
+		attr := attributeFromTags(tags)
+		if v, ok := m.gauges[name]; ok {
+			f.Observe(v, metric.WithAttributeSet(attribute.NewSet(attr...)))
+		}
+		return nil
+	}))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *metricsClient) Count(name string, value int64, tags []string, _ float64) error {
+	counter, err := m.meter.Int64Counter(name)
+	if err != nil {
+		return err
+	}
+	attr := attributeFromTags(tags)
+	counter.Add(context.Background(), value, metric.WithAttributeSet(attribute.NewSet(attr...)))
+	return nil
+}
+
+func attributeFromTags(tags []string) []attribute.KeyValue {
```

Comment:
nit: maybe we can make this return the attribute set or even the measurement option directly